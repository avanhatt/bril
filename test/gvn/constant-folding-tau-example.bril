// Ported from http://www.cs.tau.ac.il/~msagiv/courses/pa07/lecture2-notes-update.pdf

int main() {
entry:
  z : int = const 3;
  x : int = const 1;
  y : int = const 0;
  jmp check;

check:
  zero : int = const 0;
  xgt : bool = gt x zero;
  br xgt loop end;

loop:
  one : int = const 1;
  xisone : bool = eq x one;
  br xisone ifb elseb;

ifb:
  y : int = const 7;
  jmp afterb;

elseb:
  four : int = const 4;
  y : int = add z four;
  jmp afterb;

afterb:
  x : int = const 3;
  print y;
  jmp check;

end: 
  ret;
}