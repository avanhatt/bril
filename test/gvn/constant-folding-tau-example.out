int main() {
entry:
  x1: int = const 1;
  jmp copy1;
check:
  zero1: int = const 0;
  xgt1: bool = gt x2 zero1;
  br xgt1 loop end;
loop:
  one1: int = const 1;
  xisone1: bool = eq x2 one1;
  br xisone1 ifb elseb;
ifb:
  y3: int = const 7;
  jmp copy3;
elseb:
  y5: int = const 7;
  jmp copy4;
afterb:
  x3: int = const 3;
  print y4;
  jmp copy2;
end:
  ret ;
copy1:
  x2: int = id x1;
  jmp check;
copy2:
  x2: int = id x3;
  jmp check;
copy3:
  y4: int = id y3;
  jmp afterb;
copy4:
  y4: int = id y5;
  jmp afterb;
}
